go-controller/pkg/ovndbmanager/ovndbmanager.go

```diff
@@ -0,0 +1,214 @@
+		return
+	}
+	if len(out) < 4 {
+		klog.Errorf("Invalid db id found: %s for db: %s", out, db)
+		return
+	}
+	// server ID in raft membership is only first 4 char prefix
+	serverID := out[:4]
+	out, stderr, err = appCtl("cluster/status", dbName)
+	if err != nil {
+		klog.Warningf("Unable to get cluster status for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	r, _ := regexp.Compile(`Address: *((ssl|tcp):[?[a-z0-9.:]+]?)`)
+	matches := r.FindStringSubmatch(out)
+	if len(matches) < 2 {
+		klog.Warningf("Unable to parse Address for db: %s, output: %s", db, out)
+		return
+	}
+	addr := matches[1]
+	// look for current servers in raft cluster with the same address
+	r, _ = regexp.Compile("([a-z0-9]{4}) at " + addr)
+	members := r.FindAllStringSubmatch(out, -1)
+	for _, member := range members {
+		if len(member) < 2 {
+			klog.Warningf("Unable to find server id submatch in %s", member)
+			return
+		}
+		if member[1] != serverID {
+			// stale entry found for this node with same adddress, need to kick
+			klog.Infof("Previous stale member found: %s...kicking", member[1])
+			_, stderr, err = appCtl("cluster/kick", dbName, member[1])
+			if err != nil {
+				klog.Errorf("Error while kicking old Raft member: %s, for address: %s in db: %s,"+
+					"stderr: %v, error: %v", member[1], addr, db, stderr, err)
+			}
+		}
+	}
+}
+
+// ensureClusterRaftMembership ensures there are no unknown members in the current Raft cluster
+func ensureClusterRaftMembership(db string) {
+	var knownMembers, knownServers []string
+
+	var dbName string
+	var appCtl func(args ...string) (string, string, error)
+
+	if strings.Contains(db, "ovnnb") {
+		dbName = "OVN_Northbound"
+		appCtl = util.RunOVNNBAppCtl
+		knownMembers = strings.Split(config.OvnNorth.Address, ",")
+	} else {
+		dbName = "OVN_Southbound"
+		appCtl = util.RunOVNSBAppCtl
+		knownMembers = strings.Split(config.OvnSouth.Address, ",")
+	}
+	for _, knownMember := range knownMembers {
+		server := strings.Split(knownMember, ":")
+		if len(server) < 3 {
+			klog.Warningf("Failed to parse known member: %s", knownMember)
+			continue
+		}
+		knownServers = append(knownServers, server[1])
+	}
+	out, stderr, err := appCtl("cluster/status", dbName)
+	if err != nil {
+		klog.Warningf("Unable to get cluster status for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	r, _ := regexp.Compile(`([a-z0-9]{4}) at ((ssl|tcp):\[?[a-z0-9.:]+\]?)`)
+	members := r.FindAllStringSubmatch(out, -1)
+	for _, member := range members {
+		if len(member) < 3 {
+			klog.Warningf("Unable to find parse member: %s", member)
+			return
+		}
+		matchedServer := strings.Split(member[2], ":")
+		if len(matchedServer) < 3 {
+			klog.Warningf("Unable to parse address portion of the member entry: %s", matchedServer)
+			return
+		}
+		memberFound := false
+		for _, knownServer := range knownServers {
+			if knownServer == matchedServer[1] {
+				memberFound = true
+				break
+			}
+		}
+		if !memberFound && len(members) > 3 {
```

Comment:
len(members) would change after you kick off the unknown member.

Reply:
We aren't modifying the `members` slice after we get it from the regex match inside the for loop. So `len(members)` won't change as such. Plus we are predicating the `cluster/kick` firstly on `!memberFound`.

Reply:
Sorry that I wasn't clear. You have the comment "unknown member and we have enough members ...". But in case you have more than one unknown nubmers, and the len(members) is initially 4, in the end, you will kick all the unknown numbers, then at that time, the members left will no longer be enough.

Reply:
But if it's a member that is not supposed to be part of the cluster, we still need to evict that member right?

Reply:
I think we don't need that "len(members) > 3" check.

Reply:
Using the kickedMembersCount to guide the check.